test/shared/src/main/scala/zio/test/Gen.scala

```diff
@@ -436,14 +451,43 @@ object Gen extends GenZIO with FunctionVariants with TimeVariants {
     uniform.map(n => -math.log(1 - n))
 
   /**
-   * Constructs a deterministic generator that only generates the specified
-   * fixed values.
+   * Constructs a deterministic generator that generates values from the
+   * specified iterable. For infinite iterables, it will generate values from a
+   * bounded prefix determined by the current size parameter.
    */
   def fromIterable[R, A](
     as: Iterable[A],
     shrinker: A => ZStream[R, Nothing, A] = defaultShrinker
   )(implicit trace: Trace): Gen[R, A] =
-    Gen(ZStream.fromIterable(as).map(a => Sample.unfold(a)(a => (a, shrinker(a)))))
+    if (as.isEmpty) empty
+    else
+      Gen {
+        // Check if the iterable is infinite by trying to convert to Vector
+        val asVec =
+          try {
+            Some(as.toVector)
+          } catch {
+            case _: OutOfMemoryError => None
+          }
```

Comment:
I'm sorry but we really can't do this. Besides making things extremely slow as we'll need to generate collections of several GBs in size, catching an OOM Error should never _ever_ be used as an expected condition